iroh/src/sync.rs

```diff
@@ -0,0 +1,218 @@
+    debug!("sync with peer {}: finish {:?}", peer_id, res);
+    res
+}
+
+/// Runs the initiator side of the sync protocol.
+pub async fn run_alice<R: AsyncRead + Unpin, W: AsyncWrite + Unpin>(
+    writer: &mut W,
+    reader: &mut R,
+    alice: &Replica,
+) -> Result<()> {
+    let mut buffer = BytesMut::with_capacity(1024);
+
+    // Init message
+
+    let init_message = Message::Init {
+        namespace: alice.namespace(),
+        message: alice.sync_initial_message(),
+    };
+    let msg_bytes = postcard::to_stdvec(&init_message)?;
+    iroh_bytes::protocol::write_lp(writer, &msg_bytes).await?;
+
+    // Sync message loop
+
+    while let Some(read) = iroh_bytes::protocol::read_lp(&mut *reader, &mut buffer).await? {
+        debug!("read {}", read.len());
+        let msg = postcard::from_bytes(&read)?;
+        match msg {
+            Message::Init { .. } => {
+                bail!("unexpected message: init");
+            }
+            Message::Sync(msg) => {
+                if let Some(msg) = alice.sync_process_message(msg) {
+                    send_sync_message(writer, msg).await?;
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Handle an iroh-sync connection and sync all shared documents in the replica store.
+pub async fn handle_connection(
+    connecting: quinn::Connecting,
+    replica_store: ReplicaStore,
+) -> Result<()> {
+    let connection = connecting.await?;
+    debug!("> connection established!");
+    let (mut send_stream, mut recv_stream) = connection.accept_bi().await?;
+
+    run_bob(&mut send_stream, &mut recv_stream, replica_store).await?;
+    send_stream.finish().await?;
+
+    debug!("done");
+
+    Ok(())
+}
+
+/// Runs the receiver side of the sync protocol.
+pub async fn run_bob<R: AsyncRead + Unpin, W: AsyncWrite + Unpin>(
+    writer: &mut W,
+    reader: &mut R,
+    replica_store: ReplicaStore,
+) -> Result<()> {
+    let mut buffer = BytesMut::with_capacity(1024);
+
+    let mut replica = None;
+    while let Some(read) = iroh_bytes::protocol::read_lp(&mut *reader, &mut buffer).await? {
+        debug!("read {}", read.len());
+        let msg = postcard::from_bytes(&read)?;
+
+        match msg {
+            Message::Init { namespace, message } => {
+                ensure!(replica.is_none(), "double init message");
+
+                match replica_store.get_replica(&namespace) {
+                    Some(r) => {
+                        debug!("starting sync for {}", namespace);
+                        if let Some(msg) = r.sync_process_message(message) {
+                            send_sync_message(writer, msg).await?;
+                        } else {
+                            break;
+                        }
+                        replica = Some(r);
+                    }
+                    None => {
+                        // TODO: this should be possible.
```

Comment:
No it shouldn't, the receiver needs to setup the namespace/doc first, even if empty, to sync it.